crates/aptos/src/move_tool/mod.rs

```diff
@@ -1223,130 +1214,134 @@ impl Display for FunctionArgType {
-                    }),
-                    // Note commas cannot be put into the strings.  But, this should be a less likely case,
-                    // and the utility from having this available should be worth it.
-                    FunctionArgType::String => parse_vector_arg(arg, |arg| Ok(String::from(arg))),
-                    FunctionArgType::U8 => parse_vector_arg(arg, |arg| {
-                        u8::from_str(arg)
-                            .map_err(|err| CliError::UnableToParse("vector<u8>", err.to_string()))
-                    }),
-                    FunctionArgType::U16 => parse_vector_arg(arg, |arg| {
-                        u16::from_str(arg)
-                            .map_err(|err| CliError::UnableToParse("vector<u16>", err.to_string()))
-                    }),
-                    FunctionArgType::U32 => parse_vector_arg(arg, |arg| {
-                        u32::from_str(arg)
-                            .map_err(|err| CliError::UnableToParse("vector<u32>", err.to_string()))
-                    }),
-                    FunctionArgType::U64 => parse_vector_arg(arg, |arg| {
-                        u64::from_str(arg)
-                            .map_err(|err| CliError::UnableToParse("vector<u64>", err.to_string()))
-                    }),
-                    FunctionArgType::U128 => parse_vector_arg(arg, |arg| {
-                        u128::from_str(arg)
-                            .map_err(|err| CliError::UnableToParse("vector<u128>", err.to_string()))
-                    }),
-                    FunctionArgType::U256 => parse_vector_arg(arg, |arg| {
-                        U256::from_str(arg)
-                            .map_err(|err| CliError::UnableToParse("vector<u256>", err.to_string()))
-                    }),
-                    vector_type => {
-                        panic!("Unsupported vector type vector<{}>", vector_type)
-                    },
-                }?;
-                Ok(parsed)
+            serde_json::Value::Null => {
+                Err(CliError::CommandArgumentError("Null argument".to_string()))
             },
+            serde_json::Value::Object(_) => Err(CliError::CommandArgumentError(
+                "JSON object argument".to_string(),
+            )),
         }
-        .map_err(|err| CliError::BCS("arg", err))
     }
 }
 
-fn parse_vector_arg<T: Serialize, F: Fn(&str) -> CliTypedResult<T>>(
-    args: &str,
-    parse: F,
-) -> CliTypedResult<Vec<u8>> {
-    let mut parsed_args = vec![];
-    let args = args.split(',');
-    for arg in args {
-        parsed_args.push(parse(arg)?);
+// TODO use from move_binary_format::file_format_common if it is made public.
+fn write_u64_as_uleb128(binary: &mut Vec<u8>, mut val: usize) {
+    loop {
+        let cur = val & 0x7F;
+        if cur != val {
+            binary.push((cur | 0x80) as u8);
+            val >>= 7;
+        } else {
+            binary.push(cur as u8);
+            break;
+        }
     }
```

Comment:
Is there a reason we're rewriting the BCS encoding logic?

Should be able to just put it in a vector and then BCS encode it.

Reply:
> Should be able to just put it in a vector and then BCS encode it.

```rust
            serde_json::Value::Array(_) => {
                // Vector of sub-argument BCS representations.
                let mut sub_args: Vec<Vec<u8>> = vec![]; // <------------------------------------- 1
                let mut common_sub_arg_depth = None;
                // Loop over all of the vector's sub-arguments, which may also be vectors:
                for sub_arg in arg.as_array().unwrap() {
                    let ArgWithType {
                        _ty: _,
                        _vector_depth: sub_arg_depth,
                        arg: sub_arg_bcs,
                    } = self.parse_arg_json(sub_arg)?;
                    // Verify all sub-arguments have same depth.
                    if let Some(check_depth) = common_sub_arg_depth {
                        if check_depth != sub_arg_depth {
                            return Err(CliError::CommandArgumentError(
                                "Variable vector depth".to_string(),
                            ));
                        }
                    };
                    common_sub_arg_depth = Some(sub_arg_depth);
                    // Push back sub-argument BCS.
                    sub_args.push(sub_arg_bcs); // <---------------------------------------------- 2
                }
                // Default sub-argument depth is 0 for when no sub-arguments were looped over.
                Ok(ArgWithType {
                    _ty: self.clone(),
                    _vector_depth: common_sub_arg_depth.unwrap_or(0) + 1,
                    arg: bcs::to_bytes(&sub_args).map_err(|err| CliError::BCS("arg", err))?,
                    // ^ ------------------------------------------------------------------------- 3
                })
            },
            serde_json::Value::Null => {
                Err(CliError::CommandArgumentError("Null argument".to_string()))
            },
            serde_json::Value::Object(_) => Err(CliError::CommandArgumentError(
                "JSON object argument".to_string(),
            )),
```

At first glance it might appear that the above algorithm does what you are asking:

1. Declare a vector of sub-argument BCS encodings (`Vec<Vec<u8>>`)
2. Push all sub-argument BCS encodings onto the vector
3. Convert the entire `Vec<Vec<u8>>` to BCS

However, the problem with this approach is that the BCS encoding logic generates not only a `uleb128` for the top-level vector, but also for each nested vector within. This shouldn't happen, however, because the relevant BCS `Vec<u8>` encoding of each sub-argument has already been calculated (including with the leading `uleb128`) during recursion. The problem here is that when the BCS  encoder parses `Vec<Vec<u8>>` it doesn't treat each subvector `Vec<u8>` as already being in BCS form.

The reason I've rewritten the BCS logic is because it is not publicly available. Note that the same thing is already happening in [`transaction_arg_validation.rs`](https://github.com/aptos-labs/aptos-core/blob/6aed685d6c5288fa039a9ce59908fd7809d861c1/aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs#L374)